lms/djangoapps/verify_student/models.py

```diff
@@ -1037,29 +1035,30 @@ def set_deadline(cls, course_key, deadline, is_explicit=False):
                 record.save()
 
     @classmethod
-    def deadlines_for_courses(cls, course_keys):
+    def deadlines_for_enrolled_courses(cls, user):
         """
-        Retrieve verification deadlines for particular courses.
+        Retrieve verification deadlines for a user's enrolled courses.
 
         Arguments:
-            course_keys (list): List of `CourseKey`s.
+            user (user): User
 
         Returns:
             dict: Map of course keys to datetimes (verification deadlines)
 
         """
-        all_deadlines = cache.get(cls.ALL_DEADLINES_CACHE_KEY)
-        if all_deadlines is None:
-            all_deadlines = {
-                deadline.course_key: deadline.deadline
-                for deadline in VerificationDeadline.objects.all()
-            }
-            cache.set(cls.ALL_DEADLINES_CACHE_KEY, all_deadlines)
-
+        verification_deadlines_qs = VerificationDeadline.objects.extra(
+            tables=['student_courseenrollment'],
+            where=[
+                (
+                    u'course_key = student_courseenrollment.course_id and '
+                    u'student_courseenrollment.user_id = %s'
+                )
+            ],
+            params=[user.id]
```

Comment:
Erm... do you really need an `extra` here? You might be able to get away with a `VerificationDeadline.objects.filter(course_id__in=CourseEnrollment.objects.filter(user=user).values_list('id'))`

Reply:
I really want the join instead of the `select in...` if at all possible. It made a massive difference in performance for #20547, and while most users would only have a small number, we do have people who are enrolled in hundreds or thousands.

Reply:
Makes sense. I assumed the optimizer would be clever enough to figure it out with a subquery. Looking at that other PR, I notice that it's doing `list(CourseOverview.objects.values_list(...))`, which django wouldn't be able to make a subquery out of. It might be worth just experimenting in the read-replica using a user w/ a large number of courses to see if the optimizer can do the right thing w/ a subquery.

Reply:
I'll try it out and see what happens.

Reply:
You're right, it is smart enough to do something reasonable with a subquery. I'll have those changes up in a min.

Reply:
@cpennington: Change made. Can you take another look pls?